### PR TITLE
Add Security tools : Strict Cookies, Headers, and PHPStan custom rule

### LINKED ISSUE #80: [V2][Security] Implement Phpstan verification for voters usage
## Description
We need to implement a verification process using Phpstan for the usage of Symfony's Voter component

## Priority

### PR DESCRIPTION
**Pull request by nguyenk**
Fixes #80, #81, #82

### CODE REVIEW

File: apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        if ($this->functionAttributesContain(IKnowWhatImDoingThisIsAPublicRoute::class, $attributes)) {
+            return [];
+        }
+
+        $isGrantedAttribute =  $this->getAttribute(IsGranted::class, $attributes);
+
+        // LVL 1 NO PROTECTION AT ALL :: Missing vertical controls : no IsGranted and no denyAccessUnlessGranted (even without subject)
+        if ($isGrantedAttribute === null) {
```
Comment (Ngob) [inline apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php]:
```
 if ($isGrantedAttribute === null && ! $this->isDenyAccessUnlessGrantedCalledInRouteFunction($node, false)) {
 return $this->buildError(
                    sprintf('🛑🔓 SECURITY: Route %s::%s is public !', $className, $functionName),
                    "Add an #[IsGranted] attribute or use the `denyAccessUnlessGranted` function.
                    If you are sure that this route should remain public, add the " . IKnowWhatImDoingThisIsAPublicRoute::class . ' attribute',
                );
 }
```

File: apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                );
+            }
+        }
+
+        if ($this->functionAttributesContain(ThisRouteDoesntNeedAVoter::class, $attributes)){
+            return [];
+        }
+
+        // LVL 2 VERTICAL ACCESS ONLY :: IsGranted is present BUT no voter is called
+        if ($isGrantedAttribute === null) {
```
Comment (Ngob) [inline apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php]:
```
  if ($isGrantedAttribute === null && ! $this->isDenyAccessUnlessGrantedCalledInRouteFunction($node, true)) {
  return $this->buildError(
                    sprintf('🛑🔓 SECURITY: Route %s::%s is insufficiently protected !', $className, $functionName),
                    "Pass the 'subject' argument to the \$this->denyAccessUnlessGranted() call.
                    If you are sure that this route's protection should only on user's permissions, add a ".ThisRouteDoesntNeedAVoter::class." attribute.",
                );
  }
   $isGrantedAttributeInstance = $isGrantedAttribute->newInstance();
   \assert($isGrantedAttributeInstance instanceof IsGranted);
     if ($isGrantedAttributeInstance->subject === null) {
       return $this->buildError(
                    sprintf('🛑🔓 SECURITY: Route %s::%s is insufficiently protected !', $className, $functionName),
                    "Pass the 'subject' argument to the 'IsGranted' attribute.
                    If you are sure that this route's protection should only on user's permissions, add a ".ThisRouteDoesntNeedAVoter::class.' attribute.',
                );
     }   
     return [];      
  ```

File: apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        $visitor = new class ($requireSubject) extends NodeVisitorAbstract {
+            private bool $isSecurityCheckFunctionCalled = false;
+
+            public function __construct(private readonly bool $requireSubject)
+            {
+            }
+
+            public function enterNode(Node $node): int|null
+            {
+                // phpcs:disable SlevomatCodingStandard.ControlStructures.EarlyExit.EarlyExitNotUsed
```
Comment (Ngob) [inline apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php]: it should work with this CI rule with the current form

File: apps/back/src/Security/Voter/UserVoter.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    private function canCreate(): bool
+    {
+        return $this->security->isGranted(Right::ROLE_RIGHT_USER_CREATE->value);
+    }
+
+    private function canEdit(mixed $subject): bool
+    {
+        if (! $subject instanceof User) {
+            throw new \RuntimeException('Subject of voter should be a ' . User::class);
```
Comment (Ngob) [inline apps/back/src/Security/Voter/UserVoter.php:60]: not sure about the exception type
May have to create a custom ForbiddenAccessException ?

File: apps/back/config/packages/nelmio_security.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    #        redirect_status_code: 302 # default, switch to 301 for permanent redirects
+
+    # flexible HTTPS handling, read the detailed config info
+    # and make sure you have SSL working on your site before enabling this
+
+
+when@prod:
+    nelmio_security:
+        # could also be added at the proxy level
+        forced_ssl:
```
Comment (Ngob) [inline apps/back/config/packages/nelmio_security.yaml]: pas sûr que cette mecanique marche avec docker, le forced_ssl devrait être fait au niveau du reverse proxy (le reverse proxy communique en HTTP avec l'API)

File: apps/back/config/packages/framework.yaml [hunk 6]
```diff
@@ -9,7 +9,7 @@ framework:
     session:
         handler_id: null
         cookie_secure: auto
-        cookie_samesite: lax
+        cookie_samesite: strict
         storage_factory_id: session.storage.factory.native
```
Comment (Ngob) [inline apps/back/config/packages/framework.yaml]: may have to add domain explicitly (on the BASE_DOMAIN)

File: apps/back/src/Security/Voter/UserVoter.php (hunk 4 above)
Comment (nguyenk) [reply apps/back/src/Security/Voter/UserVoter.php:60]: `ForbiddenAccess` should be used if there is a permission issue no ? In this case, we are facing a `5XX` issue as `canEdit` should always be called with a `User`.

File: apps/back/config/packages/framework.yaml (hunk 6 above)
Comment (nguyenk) [reply apps/back/config/packages/framework.yaml]: Same site strict includes sub-domains so I guess we're good here

File: apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php (hunk 3 above)
Comment (nguyenk) [reply apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php]: Indeed :)

File: apps/back/config/packages/framework.yaml (hunk 6 above)
Comment (Ngob) [reply apps/back/config/packages/framework.yaml]: wont work if we allow cors from the configuration

File: apps/back/config/packages/nelmio_security.yaml [hunk 7]
```diff
@@ -0,0 +1,82 @@
+# config/packages/nelmio_security.yaml
+nelmio_security:
+    # signs/verifies all cookies
+    # FIXME :
+    #   - PHPSESSID cookie i snot correctly passed into the signing process, so we cannot use '*'. But additional cookies could be set a signed here
```
Comment (homersimpsons) [inline apps/back/config/packages/nelmio_security.yaml]: nit: wording

File: apps/back/config/packages/nelmio_security.yaml [hunk 8]
```diff
@@ -0,0 +1,82 @@
+# config/packages/nelmio_security.yaml
+nelmio_security:
+    # signs/verifies all cookies
+    # FIXME :
```
Comment (homersimpsons) [inline apps/back/config/packages/nelmio_security.yaml:4]: Can we address this now ? If no, is there any plan to address this later ?

File: apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        if ($this->functionAttributesContain(ThisRouteDoesntNeedAVoter::class, $attributes)){
+            return [];
+        }
+
+        // LVL 2 VERTICAL ACCESS ONLY :: IsGranted is present BUT no voter is called
+        if ($isGrantedAttribute === null && ! $this->isDenyAccessUnlessGrantedCalledInRouteFunction($node, true)) {
+            return $this->buildError(
+                sprintf('🛑🔓 SECURITY: Route %s::%s is insufficiently protected !', $className, $functionName),
+                "Pass the 'subject' argument to the \$this->denyAccessUnlessGranted() call.
+                    If you are sure that this route's protection should only on user's permissions, add a ".ThisRouteDoesntNeedAVoter::class." attribute.",
```
Comment (homersimpsons) [inline apps/back/src/DevTools/PHPStan/RouteSecurityChecker.php:81]: nit: wording

File: apps/back/src/DevTools/PHPStan/IKnowWhatImDoingThisIsAPublicRoute.php [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * Add this attribute to Routes that can be accessed by ANY ONE
+ * This is generally the case for
+ *   - public website / content
+ *   - Login / Password recovery
+ *   - If you are securing your route elsewhere (firewall, security.yml, custom function, etc.)
+ */
+#[Attribute(Attribute::TARGET_METHOD)]
+class IKnowWhatImDoingThisIsAPublicRoute
```
Comment (homersimpsons) [inline apps/back/src/DevTools/PHPStan/IKnowWhatImDoingThisIsAPublicRoute.php:17]: Maybe we could go with shorter class name such as `UsecurePublicRoute`

File: apps/back/src/DevTools/PHPStan/ThisRouteDoesntNeedAVoter.php [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+use Attribute;
+
+/**
+ * Add this attribute if the Route is limited to some users with specific permissions, but no horizontal check is needed
+ *  - no "ownership"
+ *  - no user context should allow / deny user from accessing the underlying resources
+ */
+#[Attribute(Attribute::TARGET_METHOD)]
+class ThisRouteDoesntNeedAVoter
```
Comment (homersimpsons) [inline apps/back/src/DevTools/PHPStan/ThisRouteDoesntNeedAVoter.php:15]: Maybe we can go with a shorter class name: `UnsecureNoVoter`

File: apps/back/config/packages/nelmio_security.yaml [hunk 12]
```diff
@@ -0,0 +1,82 @@
+# config/packages/nelmio_security.yaml
```
Comment (homersimpsons) [inline apps/back/config/packages/nelmio_security.yaml:1]: I'm not sure about all the settings in this file

**Pull request merged**